### PR TITLE
Bugfix for --make-charset with --format

### PR DESCRIPTION
**Pull request by magnumripper**
If `--format` option is used, only such pot entries will be used. This includes format groups, lists and wildcards so we can now create a charset from just unix passwords with eg. `--make-charset=FILE --format=*crypt`.

Like before, `--make-charset` without any specified format will load any pot lines including fake ones made from a wordlist with eg. `sed s/^/:/`.

As a separate commit, a KISS "cache" for the format enumeration. It sped up my test case (using `--show` with 10M mixed entries in pot file, and not specifying format) from well over a minute to just above 5 seconds. I also tried more elaborate tricks but they added very little additional performance so was ditched for the higher complexity.

### CODE REVIEW

File: src/loader.c [hunk 1]
```diff
@@ -2165,27 +2165,23 @@ static void ldr_show_pot_line(struct db_main *db, char *line)
...
-		if (!(options.flags & FLG_MAKECHR_CHK)) {
-			struct fmt_main *format = fmt_list;
+		struct fmt_main *format = fmt_list;
 
+		if (last_fmt && (ldr_trunc_valid(ciphertext, last_fmt) == 1))
+			format = last_fmt;
+		else
+		if (!(db->options->flags & DB_PLAINTEXTS) || (options.flags & FLG_FORMAT)) {
 			do {
 				if (ldr_trunc_valid(ciphertext, format) == 1)
```
Comment (solardiz) [inline src/loader.c]: Maybe add `format != last_fmt &&` here, since at this point we already know it's not that format.

**Review by solardiz — CHANGES_REQUESTED** [review]
I think as written this breaks useful functionality of being able to process lines with hashes not recognized as any format by the current build. What about, for example, building .chr files from a wordlist processed through `sed s/^/:/` to produce a fake pot file? We should keep this working.

**Comment by magnumripper** [comment]
> I think as written this breaks useful functionality of being able to process lines with hashes not recognized as any format by the current build. What about, for example, building .chr files from a wordlist processed through `sed s/^/:/` to produce a fake pot file? We should keep this working.

As I specifically mentioned almost literally, this still works (as long, obviously, as you do not give the `--format` option). This was not only dry-run but actually tested.

File: src/loader.c (hunk 1 above)
Comment (magnumripper) [reply src/loader.c]: Do you mean at [new] line 2186?

Comment (solardiz) [reply src/loader.c]: Yes, I mean new line 2186.

**Review by solardiz — APPROVED** [review]

**Comment by solardiz** [comment]
> > `sed s/^/:/` to produce a fake pot file? We should keep this working.
> 
> As I specifically mentioned almost literally, this still works (as long, obviously, as you do not give the `--format` option).

Sorry, missed that mention and misread the code.

I still think that `--show` of hashes not supported by the current build is broken with these changes, though. No? I think this used to work at least in core (only when strings directly matched without needing any `split()` magic, of course). IIRC, this was a reason why I had that case-insensitive hash function in core. But perhaps we already broke this in jumbo before this PR? I now also recall recently concluding and commenting in here that we could drop the case-insensitive hash function in jumbo. Need to find those GitHub issues now - maybe your optimization closes one of them.

**Comment by solardiz** [comment]
> Need to find those GitHub issues now - maybe your optimization closes one of them.

See #4430. Your optimization here is one of the things mentioned there.

**Comment by magnumripper** [comment]
> I still think that `--show` of hashes not supported by the current build is broken with these changes, though. No? I think this used to work at least in core (only when strings directly matched without needing any `split()` magic, of course).

With *these* changes? How? Perhaps the stuff you mention is and was borken - but I can't see this very commit making anything worse in that regard - everything should be same or better. I will take another look at this before merging though - there's really no hurry to merge. I'll take a nap and have a look with fresh eyes.

One thing I did try, apart for `sed s/^/:/`, was doing a `--show` without `--format` from an input file containing several formats, and it does show several formats' cracks in one go. I've had a feeling that was borken in Jumbo fo a while (possibly years) but I'm not quite sure it actually was and I'm not that eager to prove things - I'm just happy it does work now.

**Comment by magnumripper** [comment]
> I think this used to work at least in core (only when strings directly matched (...)

BTW and for the record, in preparation for this fix I isolated John proper's `ldr_show_pot_line` (which is like a 12-line function) and tried to make sure I understood it. There is no `strcmp` or `strcasecmp` in there that was lost so I can't really see what would fail. I'm not saying there aren't such things elsewhere but to the best of my knowledge, this function isn't it. But again, I will have another look once I've had a good night's sleep.

File: src/loader.c [hunk 2]
```diff
@@ -2194,9 +2190,21 @@ static void ldr_show_pot_line(struct db_main *db, char *line)
...
+			if (*pos == '\r' || *pos == '\n') *pos = 0;
+		} while (*pos++);
+
+		if (db->options->flags & DB_PLAINTEXTS) {
+			list_add(db->plaintexts, line);
+			return;
 		}
+
+		ciphertext = format->methods.split(ciphertext, 0, format);
+
```
Comment (magnumripper) [inline src/loader.c]: Now that I stare at this again... shouldn't we put lines 2196-2199 inside the 2201 conditional?

Comment (magnumripper) [reply src/loader.c]: No maybe not... at 2217 [new] we use this

**Comment by magnumripper** [comment]
> > I still think that `--show` of hashes not supported by the current build is broken with these changes, though. No? I think this used to work at least in core (only when strings directly matched without needing any `split()` magic, of course).
> 
> With _these_ changes? How? Perhaps the stuff you mention is and was borken - but I can't see this very commit making anything worse in that regard

I see the problem now - it wasn't caused by this commit but by the original one where we started calling `split`. I think I have a fix: If we don't find a format we shouldn't give up and return, but continue without calling any `split`. But only if no `--format` option was used.

File: src/loader.c (hunk 2 above)
Comment (solardiz) [reply src/loader.c]: We could probably speed things up by having a variation of `str_alloc_copy()` that would allocate/copy until CR/LF/NUL. That way, the string would only be scanned for a terminating character once (inside that function), not twice (before and inside). We'd then also need a variation of `list_add()` that would use this function. I'm not sure this is worth the complexity.

**Comment by magnumripper** [comment]
New version force-pushed, fixing the "show of hashes not supported by the current build". The "caching" is no longer a separate commit.

File: src/loader.c [hunk 3]
```diff
@@ -2165,10 +2165,36 @@ static void ldr_show_pot_line(struct db_main *db, char *line)
...
+		struct fmt_main *format = fmt_list;
+
+		if (last_fmt && (ldr_trunc_valid(ciphertext, last_fmt) == 1))
+			format = last_fmt;
+		else
+		if (!(db->options->flags & DB_PLAINTEXTS) || (options.flags & FLG_FORMAT)) {
+			do {
+				if (format != last_fmt && ldr_trunc_valid(ciphertext, format) == 1)
+					break;
+			} while((format = format->next));
```
Comment (solardiz) [inline src/loader.c]: Can as well introduce the missing whitespace before `((format = format->next));` while we're at it.

**Review by solardiz — APPROVED** [review]
Looks good to me.

**Comment by magnumripper** [comment]
I've retested `--make-charset` with or without format(s), and `--show` with or without format(s), including the "unsupported hash" case. It seems to be OK now.

**Comment by solardiz** [comment]
> `--show` of hashes not supported by the current build [...] used to work at least in core (only when strings directly matched without needing any `split()` magic, of course). IIRC, this was a reason why I had that case-insensitive hash function in core.

Thinking of this some more, I don't see how case-insensitive hashing could help there. It only helps arrive at a right comparison, but if the comparison itself is case-sensitive then there would be no match anyway (and for unrecognized formats the comparison should be case-sensitive, or else we'd have false positive matches for unrecognized formats where case in hashes matters). So it only helped along with `split()`, and perhaps only along with `FMT_SPLIT_UNIFIES_CASE`, and we're now obsoleting this in jumbo. So we could proceed (separately, not in this PR/commit) to drop the case-insensitivity of hashing as mentioned in #4430.

**Pull request merged**